backupccl: actually enable stats at end of download job

msbutler opened the pull request:
PR https://github.com/cockroachdb/cockroach/pull/116977 didn't re-enable auto stats collection for online restored tables
after the download job. This patch fixes this bug.

Fixes https://github.com/cockroachdb/cockroach/issues/119935

Release note: none

msbutler committed:
backupccl: actually enable auto stats at end of download job

PR #116977 didn't re-enable auto stats collection for online restored tables
after the download job. This patch fixes this bug.

Fixes #119935

Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/119955)
<!-- Reviewable:end -->

msbutler committed:
backupccl: write download job after publishing descriptors

Previously, the download job would be written before publishing sql descriptors
in the linking job. This meant that the download job could finish before the
descriptors were published. Since we attempt to manipulate descriptors at the
end of the download job (to re-enable auto stats jobs on restored tables), the
link job could fail if it observed that the restored descriptors were modified
before they were published.

This patch also fixes another bug where cluster online restore would write two
download jobs with the same ID. Previously, the download job would be written
at the end of the `restoreWithRetry()` call. Note that during cluster restore
we call this twice-- once for certain important system tables and again for the
main restoration data. This patch creates a single download job that will track
the downloading of all spans involved in the cluster restore.

Epic: none

Release note: none

msbutler commented on pkg/ccl/backupccl/restore_online.go:
```diff
@@ -501,14 +503,35 @@ func (r *restoreResumer) cleanupAfterDownload(
+			if err != nil {
+				return err
+			}
+			newTableDesc.AutoStatsSettings = &catpb.AutoStatsSettings{}
```
one sad part about writing the download job after publishing descriptors is that `tableDescs` contain the in-memory representation of the _published_ descriptors, so `tableDescs[0].AutoStatsSettings` will always be `false`. If we grabbed the `tableDescs` before publishing the descriptors, then `tableDescs[0].AutoStatsSettings` would have been set to whatever was in the backup. To get around this, we could make a deep copy of `tableDescs` before publishing the descriptors and pass that to `maybeWriteDownloadJob` . I don't really like this idea either.

stevendanna replied:
🤔 Copying aside for a moment:

I wonder if it would be worth asking foundations what they would think about an additional field in the stats settings that was something like `DisabledDuringRestore` or something.  We would then change some code in `func (as *AutoStatsSettings) AutoStatsCollectionEnabled() AutoStatsCollectionStatus` to additionally look at that field.  One reason it might be nice is that it could be used to drive other UX around what happens if the user changes the setting via SQL (since the tables will be online).

msbutler replied:
This is a great idea. let me ping sql queries.

msbutler replied:
Decided to stick with the approach of keeping all logic in backupccl.

stevendanna commented on pkg/ccl/backupccl/restore_job.go:
```diff
@@ -1974,6 +1976,40 @@ func clusterRestoreDuringUpgradeErr(
 		binaryVersion)
 }
 
+func (r *restoreResumer) maybeWriteDownloadJob(
```
Usually I like functions in the file that calls them when they only have one caller. But I wonder if in this case we want to keep this in restore_online.

msbutler replied:
Done.

stevendanna commented on pkg/ccl/backupccl/restore_online.go:
```diff
@@ -519,13 +504,35 @@ func (r *restoreResumer) cleanupAfterDownload(
+			// Setting the AutoStatsSetting field to nil implies the table will have
+			// the detault auto stats settings.
+			newTableDesc.AutoStatsSettings = nil
+			if enabled, ok := details.PostDownloadTableAutoStatsSettings[uint32(table.ID)]; ok {
+				newTableDesc.AutoStatsSettings = &catpb.AutoStatsSettings{Enabled: &enabled}
```
There are more settings in this proto that may be getting overwritten

msbutler replied:
oof. good point. fixed.

msbutler commented:
TFTR! 

bors r=dt
